### PR TITLE
Add z/elevation slicing and filtering to map renders

### PR DESCRIPTION
**Pull request by nyalldawson**
Implements https://github.com/qgis/QGIS-Enhancement-Proposals/issues/201, and adds support for setting z range for filtering map renders to QgsMapSettings/QgsRenderContext/QgsMapCanvas.

API only for now, and requires support to be added to the various map layer types.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@peterpetrik look ok to you?

**Comment by PeterPetrik** [comment]
havent got time today, i look tomorrow (and I do not mean in 10 minutes)

File: src/core/qgsrange.h [hunk 1]
```diff
@@ -167,6 +167,19 @@ class QgsRange
       return false;
     }
 
+    bool operator==( const QgsRange<T> &other ) const
+    {
+      return mLower == other.mLower &&
```
Comment (PeterPetrik) [inline src/core/qgsrange.h:172]: not sure if it is double do you want to use double comparison?

**Review by PeterPetrik — APPROVED** [review]
looks good to me!

File: src/core/qgsrange.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsrange.h:172]: Do you mean qgsDoubleNear?

Comment (PeterPetrik) [reply src/core/qgsrange.h:172]: y

Comment (nyalldawson) [reply src/core/qgsrange.h:173]: Done

**Pull request merged**